File: csharp/collatz-conjecture/CollatzConjecture.cs
using System;

public static class CollatzConjecture
{
    public static int Steps(int number)
    {
        if (number <= 0) throw new ArgumentException();

        return Calculate(number);
    }

    public static int Calculate(int number, int steps = 0)
    {
        if (number == 1) return steps;

        return Calculate(number.IsEven() ? number / 2 : number * 3 + 1, steps += 1);
    }
}

public static class IntExtenstions
{
    public static bool IsEven(this int number)
    {
        return number % 2 == 0;
    }
}

File: javascript/collatz-conjecture/collatz-conjecture.js
export const steps = num => {
  const steps = 0;
  return calculate(num, steps);
};

const calculate = (num, steps) => {
  //   console.log(num, steps);
  if (num === 1) {
    return steps;
  }

  steps += 1;

  if (num % 2 === 0) {
    calculate(num / 2, steps);
  } else {
    calculate(num * 3 + 1, steps);
  }
};
